### PR TITLE
Create StageProgress React component for header progress dots

### PR DESCRIPTION
**Pull request by joshlory**
Right now it isn't much of an improvement over the jQuery impl.  The markup is serialized through `React.renderToStaticMarkup` and jQuery still modifies the `header-level-123` elements by ID from clientState.js.  CSS still lives outside the component instead of being applied through inline styles.  The next iteration will move more of the component lifecycle inside the component.

### CODE REVIEW

File: dashboard/app/assets/javascripts/components/stage_progress.jsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        link: React.PropTypes.string,
+        id: React.PropTypes.number
+      })),
+      currentLevelIndex: React.PropTypes.number
+    },
+
+    render: function () {
+      var lastIndex = this.props.levels.length - 1;
+      var progressDots = this.props.levels.map(function (level, index) {
+
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/components/stage_progress.jsx:24]: Is the extra line of whitespace here intentional?

File: dashboard/app/assets/javascripts/components/stage_progress.jsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        var innerClass = 'level_link ' + level.status;
+        if (level.kind == 'unplugged') {
+          innerClass += ' unplugged_level';
+        }
+
+        var outerClass = (level.kind === 'assessment') ? 'puzzle_outer_assessment' : 'puzzle_outer_level';
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/components/stage_progress.jsx:30]: really feels like we should make all these types an enum, though might be better to do that in a separate PR

File: dashboard/app/assets/javascripts/components/stage_progress.jsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      }.bind(this));
+
+      return (
+        <div className="progress_container">
+          {progressDots}
+        </div>
+      );
+    }
+  });
+})(React);
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/components/stage_progress.jsx:51]: I wonder if we should have a separate ProgressDot component. Thoughts?

File: dashboard/app/assets/javascripts/header.js [hunk 4]
```diff
@@ -484,3 +460,37 @@ function populateProgress(scriptName) {
...
+      currentLevelIndex = index;
+    }
+
+    var status;
+    if (dashboard.clientState.queryParams('user_id')) {
+      // Show server progress only (the student's progress)
+      status = activityCssClass((serverProgress[level.id] || {}).result);
+    } else {
+      // Merge server progress with local progress
+      status = mergedActivityCssClass((serverProgress[level.id] || {}).result, clientProgress[level.id]);
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js]: Note: In my PR, I moved some of this stuff into it's own progress namespace, so we now have dashboard.progress.activityCssClass and dashboard.progress.mergedActivityCssClass. This might result in some merge conflicts.

File: dashboard/app/assets/javascripts/header.js [hunk 5]
```diff
@@ -51,36 +55,8 @@ dashboard.buildHeader = function (stageData, progressData, currentLevelId, userI
...
-    if (level.kind == 'unplugged') {
-      link.addClass('unplugged_level');
-    }
-    var div = $('<div>').addClass(level.id === currentLevelId ? 'puzzle_outer_current' : defaultClass).append(link);
-    if (index === levels.length - 1) {
-      div.addClass('last');
-    }
-    progressContainer.append(div).append('\n');
-  });
+  renderStageProgress(stageData, progressData, clientProgress, currentLevelId);
```
Comment (Bjvanminnen) [inline dashboard/app/assets/javascripts/header.js]: renderStageProgress might also end up making more sense living on the new namespace dashboard.progress?

File: dashboard/app/assets/javascripts/components/stage_progress.jsx [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        outerClass = (index === this.props.currentLevelIndex) ? 'puzzle_outer_current' : outerClass;
+        if (index === lastIndex) {
+          outerClass += ' last';
+        }
+
+        return ([
+          <div className={outerClass}>
+            <a id={'header-level-' + level.id} href={level.link} className={innerClass}>{level.title}</a>
+          </div>,
+          ' '
```
Comment (joshlory) [inline dashboard/app/assets/javascripts/components/stage_progress.jsx:40]: Note: The current layout of the progress dots depends on spacing between divs, e.g. `<div></div> <div></div>` instead of `<div></div><div></div>`.  I'd like to replace this with margin in the future, but opting for a pixel-equivalent layout for now.

File: dashboard/app/assets/javascripts/components/stage_progress.jsx (hunk 2 above)
Comment (joshlory) [inline dashboard/app/assets/javascripts/components/stage_progress.jsx:30]: Agreed, separate PR sounds good.

File: dashboard/app/assets/javascripts/components/stage_progress.jsx (hunk 3 above)
Comment (joshlory) [inline dashboard/app/assets/javascripts/components/stage_progress.jsx:51]: If I need to consume ProgressDot differently in the course overview I can make it a separate component then.  Otherwise if this is the only usage I don't think there's enough there to merit a new component.
